packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/ArbTokenBridge.sol

```diff
@@ -86,35 +93,70 @@ contract ArbTokenBridge is CloneFactory {
         l1Pair = _l1Pair;
     }
 
+    function mintAndCall(
+        IArbToken token,
+        uint256 amount,
+        address sender,
+        address dest,
+        bytes memory data
+    ) public {
+        // TODO: should this only be callable by mintERC20?
+        require(msg.sender == address(this), "Mint can only be called by self");
+
+        token.bridgeMint(dest, amount, '');
+        bool success = ITransferReceiver(dest).onTokenTransfer(sender, amount, data);
+
+        require(success, "External onTokenTransfer reverted");
+    }
+
     function mintERC777FromL1(
         address l1ERC20,
-        address account,
+        address sender,
+        address dest,
         uint256 amount,
-        uint8 decimals
+        uint8 decimals,
+        bytes calldata callHookData
     ) external onlyEthPair {
         IArbToken token = ensureERC777TokenExists(l1ERC20, decimals);
-        token.bridgeMint(account, amount);
+        token.bridgeMint(dest, amount, callHookData);
+        // TODO: should we trigger mintAndCall or can 777's hook deal with it?
     }
 
     function mintERC20FromL1(
         address l1ERC20,
-        address account,
+        address sender,
+        address dest,
         uint256 amount,
-        uint8 decimals
+        uint8 decimals,
+        bytes calldata callHookData
     ) external onlyEthPair {
         IArbToken token = ensureERC20TokenExists(l1ERC20, decimals);
-        token.bridgeMint(account, amount);
+
+        if(callHookData.length > 0) {
+            try ArbTokenBridge(this).mintAndCall(token, amount, sender, dest, callHookData) {
+                emit MintAndCallTriggered(true);
+            } catch {
+                // if reverted, then credit sender's account
+                token.bridgeMint(sender, amount, '');
+                emit MintAndCallTriggered(false);
+            }
+        } else {
+            // if no hook data, credit the dest
+            token.bridgeMint(dest, amount, '');
+        }
     }
 
-    function mintCustomtokenFromL1(
+    function mintCustomTokenFromL1(
         address l1ERC20,
         address account,
-        uint256 amount
+        uint256 amount,
+        bytes calldata callHookData
     ) external onlyEthPair {
         address tokenAddress = customToken[l1ERC20];
         require(tokenAddress != address(0), "Custom Token doesn't exist");
         IArbToken token = IArbToken(tokenAddress);
-        token.bridgeMint(account, amount);
+        token.bridgeMint(account, amount, callHookData);
+        // TODO: should we add postMintCall? custom token logic could implement it in bridgeMint
```

Comment:
It can't implement it without knowing the sender though

Reply:
they can encode it in the callHookData, just not as efficient

Reply:
That doesn't work if the sender is an EOA right? To rely on that it would have to come from an L1 trusted contract or something like that

Reply:
Fair, added in the try catch